ambuda/views/proofing/project.py

```diff
@@ -326,61 +319,90 @@ def search(slug):
     )
 
 
-def _replace_text(project_, replace_form: ReplaceForm, query: str, replace: str):
-    """
-    Gather all matches for the "query" string and pair them the "replace" string.
-    """
+@dataclass
+class Replacement:
+    """Models a replacement of a specific page/line."""
+
+    #: slug of the page that contains this replacement.
+    page_slug: str
+    #: The 0-indexed line number.
+    line_num: int
+    #: Regex splits of a line. Odd indices match the query phrase.
+    splits: list[str]
+    #: The replacement for the odd indices in `splits`.
+    replacement: str
+
+    @property
+    def form_key(self) -> str:
+        return f"match{self.page_slug}-{self.line_num}"
+
+    @property
+    def marked_query(self) -> str:
+        buf = []
+        for i, t in enumerate(self.splits):
+            if i % 2 == 1:
+                buf.append("<mark>")
+                buf.append(escape(t))
+                buf.append("</mark>")
+            else:
+                buf.append(escape(t))
+        return "".join(buf)
+
+    @property
+    def result(self) -> str:
+        buf = []
+        for i, t in enumerate(self.splits):
+            if i % 2 == 1:
+                buf.append(self.replacement)
+            else:
+                buf.append(t)
+        return "".join(buf)
+
+    @property
+    def marked_replace(self) -> str:
+        buf = []
+        for i, t in enumerate(self.splits):
+            if i % 2 == 1:
```

Comment:
This odd/even heuristic may not always work. I believe what you are saying is `query` is a delimiter and the resulting `splits` will always have regex `matched` elements in odd indices.